File: src/main/java/com/arnas/klatrebackend/features/climbingsessions/ClimbingSessionServiceDefault.java
package com.arnas.klatrebackend.features.climbingsessions;

import com.arnas.klatrebackend.annotation.RequireGroupAccess;
import com.arnas.klatrebackend.features.auth.GroupAccessSource;
import com.arnas.klatrebackend.features.gradesystems.Grade;
import com.arnas.klatrebackend.features.gradesystems.GradeSystemRepository;
import com.arnas.klatrebackend.features.places.PlaceRepositoryInterface;
import com.arnas.klatrebackend.features.routes.RouteRepository;
import org.jetbrains.annotations.NotNull;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Stream;

@Service
public class ClimbingSessionServiceDefault implements ClimbingSessionService {

    final private ClimbingSessionRepository climbingSessionRepository;
    final private PlaceRepositoryInterface placeRepository;
    final private GradeSystemRepository gradingSystemRepository;
    final private RouteRepository routeRepository;

    public ClimbingSessionServiceDefault(
            @Autowired ClimbingSessionRepository climbingSessionRepository,
            @Autowired PlaceRepositoryInterface placeRepository,
            @Autowired GradeSystemRepository gradingSystemRepository,
            @Autowired RouteRepository routeRepository) {
        this.climbingSessionRepository = climbingSessionRepository;
        this.placeRepository = placeRepository;
        this.gradingSystemRepository = gradingSystemRepository;
        this.routeRepository = routeRepository;
    }

    @Override
    @RequireGroupAccess
    public ClimbingSession getActiveSessionByGroup(long groupId, long userId) {
        return climbingSessionRepository.getActiveSession(groupId, userId);
    }

    @Override
    @RequireGroupAccess
    public List<ClimbingSession> getPastSessionsByGroup(long groupId, long userId) {
        return climbingSessionRepository.getPastSessions(groupId, userId);
    }


    @Override
    @RequireGroupAccess
    public void uploadSession(long userId, @NotNull ClimbingSessionDTO climbingSession) {
        throw new RuntimeException("Not yet implemented");
    }

    @Override
    @RequireGroupAccess
    @NotNull
    public ClimbingSession openSession(long groupId, long placeId, long userId) {
        var activeSessionId = climbingSessionRepository.openActiveSession(userId, groupId, placeId);
        var session = climbingSessionRepository.getClimbingSessionById(activeSessionId);
        if(session == null) throw new RuntimeException("Session not found");
        return session;
    }

    @Override
    @RequireGroupAccess(resolveGroupFrom = GroupAccessSource.FROM_SESSION)
    public void closeSession(long sessionId, boolean save, long userId) {
        var session = climbingSessionRepository.getClimbingSessionById(sessionId);
        if(session == null || !session.active()) {
            throw new RuntimeException("Session not found or not active");
        }
        int rowsAffected;
        if(save) {
            rowsAffected = climbingSessionRepository.setSessionAsInactive(session.id());
        } else {
            rowsAffected = climbingSessionRepository.deleteClimbingSession(session.id());
        }
        if(rowsAffected != 1) throw new RuntimeException("Only one row should be affected");
    }

    @Override
    @NotNull
    public RouteAttemptDisplay addRouteAttempt(long activeSessionId, long userId, @NotNull RouteAttemptDTO routeAttempt) {
        var session = climbingSessionRepository.getClimbingSessionById(activeSessionId);
        if(session == null) throw new RuntimeException("Session not found");
        if(session.userId() != userId) throw new RuntimeException("User is not the owner of the session");
        if(!session.active()) throw new RuntimeException("Session is not active");
        var insertedRouteAttempt = climbingSessionRepository.addRouteAttemptToActiveSession(activeSessionId, routeAttempt);
        RouteAttemptDisplay display;
        try {
            display = routeAttemptToDisplay(insertedRouteAttempt);
        } catch (Exception e) {
            throw new RuntimeException("Error while converting route attempt to display object");
        }
        return display;
    }

    @Override
    public RouteAttempt getRouteAttemptById(long attemptId) {
        return climbingSessionRepository.getRouteAttemptById(attemptId);
    }

    @Override
    @RequireGroupAccess(resolveGroupFrom = GroupAccessSource.FROM_SESSION)
    public void removeRouteAttempt(long attemptId, long sessionId, long userId) {
        var rowsAffected = climbingSessionRepository.deleteRouteAttempt(attemptId);
        if(rowsAffected != 1) throw new RuntimeException("Only one row should be affected");
    }

    @Override
    public List<RouteAttemptDisplay> getRouteAttempts(long sessionId) {
        var routeAttempts = climbingSessionRepository.getRouteAttemptsBySessionId(sessionId);
        return routeAttempts.stream().map((routeAttempt) -> {
            try {
                return routeAttemptToDisplay(routeAttempt);
            } catch (Exception e) {
                e.printStackTrace();
                throw new RuntimeException("Error while converting route attempt to display object");
            }
        }).toList();
    }

    @Override
    public void updateRouteAttempt(long userId, RouteAttempt routeAttempt) {
        var rowsAffected = climbingSessionRepository.updateRouteAttempt(routeAttempt);
        if(rowsAffected != 1) throw new RuntimeException("Only one row should be affected");
    }


    @Override
    public List<ClimbingSessionDisplay> getPastSessions(long groupId, long userId) {
        var pastSessions = climbingSessionRepository.getPastSessions(groupId, userId);
        return pastSessions.stream().map( session -> {
            Stream<RouteAttemptDisplay> routeDisplays = session.routeAttempts().stream().map(attempt -> {
                try {
                    return routeAttemptToDisplay(attempt);
                } catch (Exception e) {
                    throw new RuntimeException(e);
                }
            });
            return new ClimbingSessionDisplay(
                    session.id(),
                    session.userId(),
                    session.placeId(),
                    session.groupId(),
                    session.timestamp(),
                    session.active(),
                    routeDisplays.toList()
            );
        }).toList();
    }

    private RouteAttemptDisplay routeAttemptToDisplay(RouteAttempt routeAttempt) throws Exception {
        var session = climbingSessionRepository.getClimbingSessionById(routeAttempt.session());
        if(session == null) throw new Exception("Session has not been opened yet");
        var place = placeRepository.getPlaceById(session.placeId());
        if(place == null) throw new Exception("Session has not a valid place");
        var grades = gradingSystemRepository.getGradesBySystemId(place.getGradingSystemId());
        var routeOpt = routeRepository.getRouteById(routeAttempt.routeId());
        if(routeOpt.isEmpty()) throw new Exception("Route not found");
        var route = routeOpt.get();
        var gradeObject = grades.stream().filter((Grade grade) -> grade.getId() == route.getGradeId()).findFirst().orElse(null);
        if(gradeObject == null) throw new Exception("Grade not found");
        return new RouteAttemptDisplay(routeAttempt.id(), routeAttempt.attempts(), routeAttempt.completed(), route.getName(),  routeAttempt.timestamp(), gradeObject.getGradeString());
    }
}

File: src/main/kotlin/com/arnas/klatrebackend/features/climbingsessions/ClimbingSessionServiceDefault.kt
package com.arnas.klatrebackend.features.climbingsessions

import com.arnas.klatrebackend.features.auth.RequireGroupAccess
import com.arnas.klatrebackend.features.auth.GroupAccessSource
import com.arnas.klatrebackend.features.gradesystems.Grade
import com.arnas.klatrebackend.features.gradesystems.GradeSystemRepositoryDefault
import com.arnas.klatrebackend.features.places.PlaceRepository
import com.arnas.klatrebackend.features.routes.RouteRepository
import org.springframework.stereotype.Service

@Service
class ClimbingSessionServiceDefault(
    private val climbingSessionRepository: ClimbingSessionRepository,
    private val placeRepository: PlaceRepository,
    private val gradingSystemRepository: GradeSystemRepositoryDefault,
    private val routeRepository: RouteRepository
) : ClimbingSessionService {

    @RequireGroupAccess
    override fun getActiveSessionByGroup(groupId: Long, userId: Long): ClimbingSession? {
        return climbingSessionRepository.getActiveSession(groupId, userId)
    }

    @RequireGroupAccess
    override fun getPastSessionsByGroup(groupId: Long, userId: Long): List<ClimbingSession> {
        return climbingSessionRepository.getPastSessions(groupId, userId)
    }

    @RequireGroupAccess
    override fun uploadSession(userId: Long, climbingSession: ClimbingSessionDTO) {
        throw RuntimeException("Not yet implemented")
    }

    @RequireGroupAccess
    override fun openSession(groupId: Long, placeId: Long, userId: Long): ClimbingSession {
        val activeSessionId = climbingSessionRepository.openActiveSession(userId, groupId, placeId)
        return climbingSessionRepository.getClimbingSessionById(activeSessionId)
            ?: throw RuntimeException("Session not found")
    }

    @RequireGroupAccess(resolveGroupFrom = GroupAccessSource.FROM_SESSION)
    override fun closeSession(sessionId: Long, save: Boolean, userId: Long) {
        val session = climbingSessionRepository.getClimbingSessionById(sessionId)
        if (session == null || !session.active) {
            throw RuntimeException("Session not found or not active")
        }
        val rowsAffected = if (save) {
            climbingSessionRepository.setSessionAsInactive(session.id)
        } else {
            climbingSessionRepository.deleteClimbingSession(session.id)
        }
        if (rowsAffected != 1) throw RuntimeException("Only one row should be affected")
    }

    override fun addRouteAttempt(activeSessionId: Long, userId: Long, routeAttempt: RouteAttemptDTO): RouteAttemptDisplay {
        val session = climbingSessionRepository.getClimbingSessionById(activeSessionId)
            ?: throw RuntimeException("Session not found")
        if (session.userId != userId) throw RuntimeException("User is not the owner of the session")
        if (!session.active) throw RuntimeException("Session is not active")
        val insertedRouteAttempt = climbingSessionRepository.addRouteAttemptToActiveSession(activeSessionId, routeAttempt)
        return try {
            routeAttemptToDisplay(insertedRouteAttempt)
        } catch (e: Exception) {
            throw RuntimeException("Error while converting route attempt to display object")
        }
    }

    override fun getRouteAttemptById(attemptId: Long): RouteAttempt {
        return climbingSessionRepository.getRouteAttemptById(attemptId)
    }

    @RequireGroupAccess(resolveGroupFrom = GroupAccessSource.FROM_SESSION)
    override fun removeRouteAttempt(attemptId: Long, sessionId: Long, userId: Long) {
        val rowsAffected = climbingSessionRepository.deleteRouteAttempt(attemptId)
        if (rowsAffected != 1) throw RuntimeException("Only one row should be affected")
    }

    override fun getRouteAttempts(sessionId: Long): List<RouteAttemptDisplay> {
        val routeAttempts = climbingSessionRepository.getRouteAttemptsBySessionId(sessionId)
        return routeAttempts.map { routeAttempt ->
            try {
                routeAttemptToDisplay(routeAttempt)
            } catch (e: Exception) {
                e.printStackTrace()
                throw RuntimeException("Error while converting route attempt to display object")
            }
        }
    }

    override fun updateRouteAttempt(userId: Long, routeAttempt: RouteAttempt) {
        val rowsAffected = climbingSessionRepository.updateRouteAttempt(routeAttempt)
        if (rowsAffected != 1) throw RuntimeException("Only one row should be affected")
    }

    override fun getPastSessions(groupId: Long, userId: Long): List<ClimbingSessionDisplay> {
        val pastSessions = climbingSessionRepository.getPastSessions(groupId, userId)
        return pastSessions.map { session ->
            val routeDisplays = session.routeAttempts.map { attempt ->
                try {
                    routeAttemptToDisplay(attempt)
                } catch (e: Exception) {
                    throw RuntimeException(e)
                }
            }
            ClimbingSessionDisplay(
                id = session.id,
                userId = session.userId,
                placeId = session.placeId,
                groupId = session.groupId,
                timestamp = session.timestamp,
                active = session.active,
                routeAttempts = routeDisplays
            )
        }
    }

    @Throws(Exception::class)
    private fun routeAttemptToDisplay(routeAttempt: RouteAttempt): RouteAttemptDisplay {
        val session = climbingSessionRepository.getClimbingSessionById(routeAttempt.session)
            ?: throw Exception("Session has not been opened yet")
        val place = placeRepository.getPlaceById(session.placeId)
            ?: throw Exception("Session has not a valid place")
        val grades = gradingSystemRepository.getGradesBySystemId(place.gradingSystemId)
        val routeOpt = routeRepository.getRouteById(routeAttempt.routeId)
        if (routeOpt.isEmpty) throw Exception("Route not found")
        val route = routeOpt.get()
        val gradeObject = grades.firstOrNull { grade: Grade -> grade.id == route.gradeId }
            ?: throw Exception("Grade not found")
        return RouteAttemptDisplay(
            id = routeAttempt.id,
            attempts = routeAttempt.attempts,
            completed = routeAttempt.completed,
            routeName = route.name,
            timestamp = routeAttempt.timestamp,
            gradeName = gradeObject.gradeString
        )
    }
}

